src/shared/boot-entry.c

```diff
@@ -0,0 +1,220 @@
+
+static int entry_token_from_machine_id(sd_id128_t machine_id, BootEntryTokenType *type, char **ret) {
+        char *p;
+
+        assert(type);
+        assert(IN_SET(*type, BOOT_ENTRY_TOKEN_AUTO, BOOT_ENTRY_TOKEN_MACHINE_ID));
+        assert(ret);
+
+        if (sd_id128_is_null(machine_id))
+                return 0;
+
+        p = strdup(SD_ID128_TO_STRING(machine_id));
+        if (!p)
+                return log_oom();
+
+        *ret = p;
+        *type = BOOT_ENTRY_TOKEN_MACHINE_ID;
+        return 1;
+}
+
+static int entry_token_from_os_release(const char *root, BootEntryTokenType *type, char **ret) {
+        _cleanup_free_ char *id = NULL, *image_id = NULL;
+        int r;
+
+        assert(type);
+        assert(IN_SET(*type, BOOT_ENTRY_TOKEN_AUTO, BOOT_ENTRY_TOKEN_OS_IMAGE_ID, BOOT_ENTRY_TOKEN_OS_ID));
+        assert(ret);
+
+        switch (*type) {
+        case BOOT_ENTRY_TOKEN_AUTO:
+                r = parse_os_release(root,
+                                     "IMAGE_ID", &image_id,
+                                     "ID",       &id);
+                break;
+
+        case BOOT_ENTRY_TOKEN_OS_IMAGE_ID:
+                r = parse_os_release(root, "IMAGE_ID", &image_id);
+                break;
+
+        case BOOT_ENTRY_TOKEN_OS_ID:
+                r = parse_os_release(root, "ID", &id);
+                break;
+
+        default:
+                assert_not_reached();
+        }
+        if (r == -ENOENT)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to load %s/etc/os-release: %m", strempty(root));
+
+        if (!isempty(image_id) && boot_entry_token_valid(image_id)) {
+                *ret = TAKE_PTR(image_id);
+                *type = BOOT_ENTRY_TOKEN_OS_IMAGE_ID;
+                return 1;
+        }
+
+        if (!isempty(id) && boot_entry_token_valid(id)) {
+                *ret = TAKE_PTR(id);
+                *type = BOOT_ENTRY_TOKEN_OS_ID;
+                return 1;
+        }
+
+        return 0;
+}
+
+int boot_entry_token_from_type(
+                const char *root,
+                const char *conf_root, /* will be prefixed with root */
+                sd_id128_t machine_id,
+                BootEntryTokenType *type,
+                char **ret) {
+
+        int r;
+
+        assert(type);
+        assert(ret);
+
+        switch (*type) {
+
+        case BOOT_ENTRY_TOKEN_AUTO:
+                r = entry_token_load(root, conf_root, type, ret);
+                if (r != 0)
+                        return r;
+
+                r = entry_token_from_machine_id(machine_id, type, ret);
+                if (r != 0)
+                        return r;
+
+                r = entry_token_from_os_release(root, type, ret);
+                if (r != 0)
+                        return r;
+
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "No machine ID set, and %s/etc/os-release carries no ID=/IMAGE_ID= fields.",
```

Comment:
Shouldn't `conf_root` be used here?

Reply:
No, 'conf_root' is typically /etc/kernel.
I will rename the variable to `etc_kernel`, like `bootctl` named.